### ISSUE TITLE
Remove unused variables from puf.csv file

### ISSUE THREAD

**Issue by martinholmer** [open-source-economics/Tax-Calculator#609]
Tax-Calculator pull request #603 adds a Records.UNUSED_READ_VARS set which contains variables that are in the Records.VALID_READ_VARS set, but are not used in the tax calculations in any way.  These unused variables do not have an array created for them in the Records call object.  It would be more efficient, and less confusing to newcomers, if these unused variables were removed from the ```puf.csv``` file.

The removing of these variables is probably best done in the taxdata repository as a new step in the final-prep/puf-cps-processing.py script.

The size reduction in the ```puf.csv``` file would be substantial as shown by the following tabulations:
```
$ python
Python 2.7.11 |Anaconda 2.3.0 (x86_64)| (default, Dec  6 2015, 18:57:58) 
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
>>> from taxcalc import Records
>>> print len(Records.VALID_READ_VARS)
228
>>> print len(Records.UNUSED_READ_VARS)
121
>>> quit()
$ 
```

So, more than half of the variables in the ```puf.csv``` file are never used in the Tax-Calculator.

**Comment by Amy-Xu** [comment]
Martin, I haven't go though the UNUSED_READ_VARS list thoroughly, but my first intuition is to keep those variables in puf.csv, and have them accessible through records class. My argument in general is that our users might want to see these variables for checking/testing purposes, or might even create tax logic with these currently unused variables. One example is variable `filer`, which documents filers and non-filers tag for 2009. Some users might want to compare any later year non-filers with 09. If we removed this variable from puf, users could not even extract this variable through our python interface, which will not be convenient for them.

What do you think?

**Comment by martinholmer** [comment]
Amy said:
> Martin, I haven't gone though the UNUSED_READ_VARS list thoroughly, but my first intuition is to keep those variables in ```puf.csv```, and have them accessible through Records class. My argument in general is that our users might want to see these variables for checking/testing purposes, or might even create tax logic with these currently unused variables. One example is variable ```filer```, which documents filers and non-filers tag for 2009. Some users might want to compare any later year non-filers with 09. If we removed this variable from ```puf.csv```, users could not even extract this variable through our Python interface, which will not be convenient for them.
>
> What do you think?

Well, you (and Matt who made a similar point on the phone) are correct in that my proposal to remove the unused variables from ```puf.csv``` has pros and cons.  Perhaps there are a few variables that could be used in tax-logic enhancements (both you and Matt offered only one: ```filer```), but my impression is that most of the unused variables are just junk.  For example, why are ```s008``` and ```s009``` in the ```puf.csv``` file?  In what kind of situation would these variables be used?

I think you and Matt are way underestimating the confusion factor with new users/contributors.  I'm an experienced user of the Python Tax-Calculator and I'm confused as to why **more than half** of the ```puf.csv``` variables are used nowhere in the tax-calculation logic.

If new users get sophisticated enough to want to program a tax reform that uses variables that are currently unused in Tax-Calculator, they will certainly be sophisticated enough to ask OSPC for an expanded version of ```puf.csv``` that includes the extra variables they need to program their reform.  And in that case, there will be no need to distribute that expanded version of ```puf.csv``` to anybody else unless this sophisticated user's new reform gets merged into the OSPC Git Hub master branch.  Remember your example assumes that this sophisticated user has paid IRS to use the 2009 SOI PUF, so they will know what is available in that dataset that is not in our pared-down ```puf.csv``` file.

So, on balance, I can see your point about including the ```filer``` variable, but I don't find that one case to be a compelling argument to keeping the other 120 unused variables in the ```puf.csv``` file.  Maybe I'm missing something?  Which of the other 120 unused variables do you think should be keep in the ```puf.csv``` file even though they are currently not being used?  And why?

@Amy-Xu @MattHJensen @feenberg @GoFroggyRun

**Comment by Amy-Xu** [comment]
Indeed I see it might be confusing for newcomers. Removing the redundant variables would also help to save some space on our Amazon nodes. But I feel we need to point our users to the taxdata file where these extra variables got removed, so if they ever want to add any of them back, they can do it easily. I only have `filer` in my mind at this point, and will take a look see whether we need to keep any others (shouldn't be too many). How does that sound?

**Comment by martinholmer** [comment]
Amy said:
> Indeed I see it might be confusing for newcomers. Removing the redundant variables would also help to save some space on our Amazon nodes. But I feel we need to point our users to the taxdata file where these extra variables got removed, so if they ever want to add any of them back, they can do it easily.

That's an excellent idea.  Because if they want to add variables that means they have paid IRS for the 2009 data and they should be able to use the Python scripts in the PUBLIC taxdata repo to construct their own expanded version of the ```puf.csv``` file.  We provide the open-source scripts, but each user must provide their own copy of the proprietary IRS SOI PUF dataset.

Amy also said:
> I only have filer in my mind at this point, and will take a look see whether we need to keep any others (shouldn't be too many). How does that sound?

I think this is the correct approach.  But after thinking about your hypothetical situation, I'm still wondering about the usefulness of the ```filer``` variable.  Earlier you said:
> One example is variable filer, which documents filers and non-filers tag for 2009. Some users might want to compare any later year non-filers with 09.

What variable in Tax-Calculator simulates whether or not a filing unit filed a tax return in 2013+?  What simulated variable would users compare to the 2009 ```filer``` variable in the ```puf.csv``` file?  I didn't immediately see what variable it would be, but I don't focus on those sort of things much.

@Amy-Xu @MattHJensen @feenberg @GoFroggyRun

**Comment by Amy-Xu** [comment]
I'm using this variable to check whether all marked non-filers are indeed having zero liabilities. I started with checking the original data in 2009 and then calculated values starting from 2013. Does it make sense?

I was thinking may be we can incorporate both this removing-extra-variable step and the other 08-var step in issue #576 in this [taxdata script](https://github.com/open-source-economics/taxdata/blob/master/final%20prep/puf-cps-processing.py). May be start with if it's 08 PUF, then abc, else if 09 PUF then def. And then remove all the unused variables. What do you think?

**Comment by martinholmer** [comment]
Sounds like a very sensible approach.

**Comment by talumbau** [comment]
@martinholmer @Amy-Xu I wanted to chime in here because the outcome of this discussion may result in a nice performance improvement for users in general, but also for TaxBrain users in particular. Let me explain.

When a user creates a `Records` object, he/she either has already created a DataFrame (using `pd.read_csv`) and is passing it into the `Records` constructor, or a filename is passed (in which case, `pd.read_csv` is called in the constructor). It turns out that there is an argument to `read_csv`, `usecols`, that allows one to specify the name of the columns that one wishes to read out of the file. The CSV reader in `pandas` is extremely good and implemented in a combination of C and Cython. By specifying the desired columns only, a lot of low-level performance optimizations are possible. The result is that one can significantly reduce the time to read in the data, if the desired columns are a small subset of the total number of columns available in the file.

It was through this issue that I learned that the ratio of "used columns" to "existing columns" in the PUF is about 1:2. So, the opportunity for speedup is there.

I ran a small experiment. We have many sets of columns in the `Records` class (e.g. `Records.UNUSED_READ_VARS`, `Records.VALID_READ_VARS`, etc.). Interestingly, I did not find a variable that covers the case of "here's everything we need to read out of the PUF." So, I created a `Records` object in the typical way, but I kept track of every column that we read in. Then, I subtracted out `Records.UNUSED_READ_VARS` to get a final list of variables. Then, I created a `Records` object out of a DataFrame built from this set of "only columns I need" variables. I timed both (the typical `Records` object creation, and the one with `usecols`) and it is about twice as fast to use the `usecols` argument on the set of "necessary columns". Here is the gist of the experiment:

https://gist.github.com/talumbau/eff073baebbff760daed

Additionally, in TaxBrain, we end up copying this raw DataFrame as we created difference `Records` objects (that each need their own copy of the data) for the dropq calculation. So, if we are copying less data in memory, that is also a good thing for performance.

So, here is my request: I believe the `Records` class should maintain a set of column names that are the columns that our calculator uses in the PUF. By explicitly have such a list of column names (perhaps something like `Records.NECESSARY_VARS`), we can then modify our `read_csv` as follows:

```
puf_df = pd.read_csv("puf.csv", usecols=Records.NECESSARY_VARS)
```

Similarly, if a user wants to create the DataFrame outside of using the Calculator, he/she can still have access to the list of names that one needs through such a constant.

If the goal is to remove columns we don't need from the PUF, my request can be satisfied simply by maintaining a list of names of the remaining columns inside the Records class. Does that make sense?

I/O and memory copying are a significant chunk of the overall compute time on TaxBrain, so this could have a very positive impact on the run times that users see. cc @MattHJensen

**Comment by Amy-Xu** [comment]
This sounds awesome!! Didn't know this step could potentially speed up the calculations. As I have already finished the script removing all unused variables, I think we can just proceed with a new version of 'PUF', and read in everything inside this file?

The new data file is 61.7M, about one half size of the previous version.

**Comment by talumbau** [comment]
Yes, that sounds good to read everything inside the file, but I still think it is best to be explicit about which columns, so I think the result of the work by Amy should be a list of the columns from the PUF that we use as a constant in the `Records` class (which should be the same as the first line of the new puf.csv that Amy has created)

**Comment by martinholmer** [comment]
No, I don't think this will work. See all the tests that show how we want people to be able to input as few as two variables.

**Comment by martinholmer** [comment]
T.J. said:
> I/O and memory copying are a significant chunk of the overall compute time on TaxBrain, so this could have a very positive impact on the run times that users see.

When I looked at your gist, I didn't see any of test results.  How much faster would a typical TaxBrain run be if we were able to do the optimization you suggest when reading the ```puf.csv``` file?  Would TaxBrain run 10% faster, 30% faster, or what?  Right now the execution time for TaxBrain is just a minute of so (or so it seems as TaxBrain winks at me).  By how much would that execution time be reduced if we were somehow able to do the optimization?

cc @talumbau @MattHJensen @feenberg @Amy-Xu @GoFroggyRun
